### DISCUSSION TITLE
Calculate savings of discharging battery minus inverter loss and cost of battery

### DISCUSSION THREAD

**Discussion by tobiasnorbo** [JensenNick/huawei_solar_pees#42] [Question / Answer] [answered]
I'm trying to implement the overview of how much I earn from discharging my battery pack, using the wiki section 7 (bottom of the page). I understand as much, that you need to make a sensor from 'Energi Data Service' that has the sales price and then return a negative value. But as I understood from the latest update there should be some smart calculations of energy loss during AC to DC conversion? Is this true for subtracting the cost of the system as well? Currently I have made a sensor letting me know, when it makes sense to sell the energy from the battery to the grid. In this case the calculation looks something like: Energy Data service price (without vat and fees) minus sales price to energy provider, minus battery pack cost price / (10000 charge cycles * 10 kwh). To have something more close to reality I have added a 10% loss to the calculation.
Do you recommend me to do the same or is there a more exact way to do it with the new sensors?
Thanks in advance

**Comment by JensenNick** [comment, marked as answer]
The PEES package includes the `sensor.battery_discharge_grid_sale`. This sensor will calculate the income you have from discharging the battery to grid. 

If you'd like this to be implemented in the Energy Dashboard, yes then you would need to add a negative price sensor in order for the amount to appear as a negative cost (you only have costs in the Energy Dashboard) according to the Wiki Pages section 7.

In the STAT package you have the `sensor.battery_efficiency_total_ts`, which will calculate the efficiency of your battery. You can use this either directly or as a guide to set a static factor. I would factor in this efficiency in your calculation and set it up as follows.
Charge only if "electrical price at the time of discharge" is greater than "electrical price at time of charge / (efficiency factor / 100) + depreciation" and I would calculate the depreciation as you do "battery cost / (total cycles * capacity)". Note that the efficiency factor you get from the sensor is in percent so you'd need to divide it with 100.

I'm working on setting an automation like this up, but just haven't got around it.

Hope that answers your question.

**Reply by tobiasnorbo** [reply]
Thanks a lot Nick!

I will try it out.

**Reply by tobiasnorbo** [reply]
Just a quick follow-up question. The sensor.battery_efficiency_total_ts seems to be dropping steadily since I implemented the STAT package on 26th of March. Here it was at around 86%, and it has been dropping steadily to now around 77,9%. Is that normal? And do you know how the sensor works/calculates the efficiency? Thanks :)

**Reply by JensenNick** [reply]
The sensor is the quotient (result) of energy discharged (kWh output AC) from the battery and energy charged onto the battery (kWh input either as DC from PV or AC from grid). I’ve explained a bit about it in the README and in the WikiPages. This is your true roundtrip efficiency. Do note that I've yet to come across a test or review that document the true roundtrip efficiency of any battery. Not even the German test standard (STC) used by Fronius and BYD among others, tell you the true roundtrip efficiency.

I monitored my battery during this winter, where I charged almost every night and discharged during the evening. My efficiency was 77%. We had almost no sun. Now we have a tad more sun and the efficiency is improving a bit so, my efficiency is around 80% at the moment. Did you set your start values according to the README?

My battery is a 10 kWh battery placed in an 18 degree garage. A 5 kWh battery will have a higher relative standby loss. The surrounding temperature will also influence the efficiency.

I’ve made several posts on the danish FB forum and a single post on the international FB forum about this - advocating to factor in the efficiency when charging.

The international FB forum:
https://www.facebook.com/groups/469094214903015/permalink/844532804025819/

The danish FB forum:
https://www.facebook.com/groups/635006747974900/permalink/900323734776532/

https://www.facebook.com/groups/635006747974900/permalink/887306516078254/

https://www.facebook.com/groups/635006747974900/permalink/875318187277087/

**Reply by tobiasnorbo** [reply]
Thank you for the elaboration. Firstly, I have a similar setup with 10kWh battery. It is placed in an isolated shed, that has been above 10 degrees all winter, and due to the inverter being in there as well, the temperature now has increased, so probably similar to your setup in terms of temperature.
I believe I have done the setup according to the README, but I couldn't find any relevant information on what to do with the battery setup - see snapshot. 
I havn't seen the posts on the different fb sites. Will give it a look.
<img width="391" alt="Screenshot 2024-04-07 at 20 46 28" src="https://github.com/JensenNick/huawei_solar_pees/assets/162321792/8e7d268d-ac6b-40aa-8cd4-3bc45514ab97">
<img width="378" alt="Screenshot 2024-04-07 at 20 47 22" src="https://github.com/JensenNick/huawei_solar_pees/assets/162321792/f5298432-d76e-4044-b71c-895d64330535">

**Reply by JensenNick** [reply]
A couple of things... 1) Is it correct that the starting values for the battery both are 0? The README explains how to set these values. 2) It looks like you have two "operating voltages pr. inverter" or is it just because you are showing all possible values? My view does not do that.
Here is what my "Input view" look like. I've added the sensors with info about the PV voltage.
![2024-04-09_09-12-11](https://github.com/JensenNick/huawei_solar_pees/assets/149193835/9c596486-b352-47e5-b7fa-47a34f44b4f5)

**Reply by tobiasnorbo** [reply]
Thank you for your time helping me out. 

The attachement was from the Overview card, that HA makes per default. That is why it showed the operating  voltage twice. I have made an input card as you suggested in your README, and deleted stuff that I don't need, since I only have one 3 phase inverter. Made sure the others was toggled off etc. before deleting.

But I have read through you README several times without finding a guide for setting the battery efficiency parameters. Have you deleted this?

Guidance here would be appreciated. It is still dropping for me. Now at 76%.

For now I will set the values similar to yours (not really possible,since the steps only allow increments of 1), but have no clue what it means and what it does. I have set them to 0 and 5 kWh, respectively.

**Reply by tobiasnorbo** [reply]
Edit: When I changed the battery efficiency settings to 0 and 5, the efficiency dropped to 73,8%.

**Comment by Roving-Ronin** [comment]
@JensenNick 

Just looking at your screenshot above and in the 'Start Value Charge (kWh)' and 'Start Value Discharge (kWh)' you have is is 0.62kWh and 4.94kWh?

Just wondering if this is correct, as given it says kWh I am trying to enter 0kWh and 4.061kWh and the input rejects it, only accepting whole number increments.

Should the below sensors read "(in W)" instead of kWh ? I am assuming so, given the max=120000 (that also seems huge in kWh, given each inverter can run 2 x 15kWh LUNA S0 = 30000W, and the new LUNA S1 can be run with 4 x 21kWh per inverter = 81000W total / 162000W for two inverters each with 4 LUNA S1 ... if you where a millionaire to afford that many).

```
  battery_efficiency_start_value_charge:
    name: "Battery Efficiency Start Value Charge (in kW)"
    min: 0
    max: 120000
    step: 1
    mode: box

  battery_efficiency_start_value_discharge:
    name: "Battery Efficiency Start Value Disharge (in kW)"
    min: 0
    max: 120000
    step: 1
    mode: box
```

**Reply by JensenNick** [reply]
@Roving-Ronin I believe you're right. Didn't notice an issue earlier, but I do see the red marking under the box and the notation now. Maybe I put the values in via Developer Tools. It'll need an revision.

**Reply by Roving-Ronin** [reply]
@JensenNick 

Instead of manually defining the:

- Battery size and/or number (and in future needing to see if it's an S0 with 5/10/15kWh in the stack or the new S1 with 7/14/21kWh in the stack)
- 1 phase or 3 phase
- How many inverters and model/size

How you considered looking for this information from WLCRS integration, and using that in conditional sensors and cards?

For example, I note that WLCRS is getting the makes and model (& even serial numbers) from the Inverters / Batteries / Power Meter / Optimisers.

This info only needs to be queried 'at boot/restart' and With this it should be possible for instance to say 'if power meter is DDSU then set 1 phase or if DTSU set 3 phase' (there's a new meter just coming out and the EMMA that is also a meter, but I'm sure you get the idea. Or else this could also be taken from the models L0 / L1 is 1p and M0 / M1 / M2 / M3 (& the new MB0 you need to add) or 3p.

WLCRS is getting this info, but it isn't published actively in a sensor... if it can be traced to where it is (I'd have to lookup the python lib it uses and the Modbus registers) it could probably be pulled out (assume it's listed as an attribute somewhere atm) and made available via a custom sensor.

Adds some coding and automated smarts into it, but also means it makes it more KISS (keep it simple stupid) for end users.

**Reply by JensenNick** [reply]
@Roving-Ronin I'm not a programmer / don't know Python or other languages so I rely on the information presented as sensors. Of course it could be cool to to have some semi automation or convert the whole package to an integration - but not something I can support at the moment.

PS: I have a DTSU and I have both a 1 phase and a 3 phase inverter + if you have a battery and a 3 phase house installation you'd need a DTSU also for a 1 phase inverter.

**Reply by Roving-Ronin** [reply]
I should mentioned a word of 'warning' I will have to update the derived Hauwei sensors yaml with the forthcoming release of the WLCRS v1.4.0 integration for Huawei Solar. This is as v1.4.0 has updated a number of the sensors and introduced some new ones, notably the introduction of querying the optional (you have to enable the sensors) "Batteries" sensors. This means we will no longer have to manually create these batteries (plural) sensors to provide sensors for total yield, active power, input power etc... instead the WLCRS / Huawei modbus will provide them.

If anyone has the manually defined batteries sensors, when the updates release, I'd suggest disabling the yaml file and restarting HA and then deleting the old 'batteries' sensors (they have the same name as the new ones, so the historical data will remain and link to the new ones), then re-enabling the derived sensors file (after updating it to the new version) and then restarting HA again, then update WLCRS from v1.3.3 to v.1.4.0
